### ISSUE TITLE
Augmentation layers only active during training!

### ISSUE THREAD

**Issue by innat** [keras-team/keras#17385]
**Description of issue (what needs to be changed):**.
**Correct links**. https://keras.io/guides/preprocessing_layers/

As it said, the following layers apply random augmentation transforms to a batch of images. They are only active **during training**.

[tf.keras.layers.RandomCrop](https://keras.io/api/layers/preprocessing_layers/image_augmentation/random_crop#randomcrop-class)
[tf.keras.layers.RandomFlip](https://keras.io/api/layers/preprocessing_layers/image_augmentation/random_flip#randomflip-class)
[tf.keras.layers.RandomTranslation](https://keras.io/api/layers/preprocessing_layers/image_augmentation/random_translation#randomtranslation-class)
[tf.keras.layers.RandomRotation](https://keras.io/api/layers/preprocessing_layers/image_augmentation/random_rotation#randomrotation-class)
[tf.keras.layers.RandomZoom](https://keras.io/api/layers/preprocessing_layers/image_augmentation/random_zoom#randomzoom-class)
[tf.keras.layers.RandomHeight](https://keras.io/api/layers/preprocessing_layers/image_augmentation/random_height#randomheight-class)
[tf.keras.layers.RandomWidth](https://keras.io/api/layers/preprocessing_layers/image_augmentation/random_width#randomwidth-class)
[tf.keras.layers.RandomContrast](https://keras.io/api/layers/preprocessing_layers/image_augmentation/random_contrast#randomcontrast-class)

So, in the model, if we use these layers, they will be **active** (i.e. `model.fit` / `model(..., training=True)` during training but will be **inactive** when we call `model.evaluate` and `model.predict` / `model(..., training=False)`. A sample model would be 

```python
model = keras.Sequential(
   [
      tf.keras.layers.RandomHeight,
      tf.keras.layers.RandomContrast,
      tf.keras.layers.Conv2D,
      tf.keras.layers.Dense
      ...
    ]
)
```  

Queries

- What if I want the augmentaiton layer be active in the prediction phase (i.e for making test-time-augmentaiton), using `model.predict`? We can use `model(..., training=True)` but (a). It's best suited for small amount of data, and (b). some other function (`dropout`) might be active. 
- Already checked it. https://github.com/keras-team/keras/blob/e6784e4302c7b8cd116b74a784f4b78d60e83c26/keras/backend.py#L451-L456

**Comment by Frightera** [comment]
> What if I want the augmentaiton layer be active in the prediction phase (i.e for making test-time-augmentaiton), using model.predict

I think one way would be writing a wrapper class and force `training=True`:

```
class WrappedRandomFlip(layers.experimental.preprocessing.PreprocessingLayer):
    def __init__(self, mode, seed):
        super(WrappedRandomFlip, self).__init__()
        self.layer = layers.RandomFlip(mode=mode, seed=seed)
    
    def call(self, inputs, training):
        return self.layer(inputs, training=True)
```

Otherwise using `training=True` can trigger such regularization layers like Dropout etc. as you said.

**Comment by innat** [comment]
@Frightera thanks, It seems a standard approach. But to make it work, I've to rewrite the model definision in the prediction time.

```python
model = [random_layers, conv2d, dense, ..]
model.fit 
model.evaluate 

model = [wrapper_random_layers, , model[1:]]
model.predict
```

**Comment by Frightera** [comment]
@innat What about inheriting `BaseImageAugmentationLayer` directly and creating an attribute called `test_time_aug`?

```python
from keras.layers.preprocessing.image_preprocessing import BaseImageAugmentationLayer

class WrappedRandomFlip(BaseImageAugmentationLayer):
    def __init__(self,
                 mode="horizontal_and_vertical",
                 seed=None):
        
        super(WrappedRandomFlip, self).__init__()
        
        self.layer = layers.RandomFlip(mode=mode, seed=seed)
        self.test_time_aug = False
    
    def call(self, inputs, training):
        if self.test_time_aug:
            return self.layer(inputs, training=True)
        else:
            return self.layer(inputs, training=training) 

random_layer = WrappedRandomFlip()                  
```

Then you need to override/wrap your predict function into another one and check the layers of the model using `isinstance(layer, BaseImageAugmentationLayer)` and set `test_time_aug = True`.

**Comment by innat** [comment]
@Frightera Thanks, really appreciate your suggestion.  I expected to get it directly without any sort of workaround. But as augmentation only supposed to be applied on training set, so that makes sense to disable it fully in the testing time. From your suggestion, here is one way to achieve. I'm also interested to know if it's the best way to do this from the API until now. (cc. @rchao)

```python
class TrainingAugmentationLayers(keras.layers.Layer):
    def __init__(self):
        super().__init__()
        self.random_flip = RandomFlip("horizontal")
        self.random_zoom = RandomZoom(0.1, fill_mode='reflect')
        self.random_rotate = RandomRotation(0.1, fill_mode='reflect')
    def call(self, inputs):
        x = self.random_flip(inputs)
        x = self.random_zoom(x)
        x = self.random_rotate(x)
        return x

class PredictAugmentaitonLayers(TrainingAugmentationLayers):
    def call(self, inputs):
        x = self.random_flip(inputs, training=True)
        x = self.random_zoom(x, training=True)
        x = self.random_rotate(x, training=True)
        return x
```
```python
img = tf.random.normal([20, 512, 512, 3], 0, 1, tf.float32)
y_true = img

inputs = tf.keras.Input(shape=(512, 512, 3))
x = TrainingAugmentationLayers()(inputs)
y = PredictAugmentaitonLayers()(inputs)
model = tf.keras.Model(inputs=inputs, outputs=[x, y])
```
```python
y_pred1, y_pred2 = model.predict(img)
np.allclose(y_pred1, y_true) # True
np.allclose(y_pred2, y_true) # False
```

**Comment by divyashreepathihalli** [comment]
@innat - It is expected behavior for augmentation to run only during training. However if you want augmented data during inference too, I have some questions
- Do you expect your model to always augment during inference? - you might not know what to expect in the results. - If you want this to be your final model behavior, you can create a wrapper class that subclasses the `random*` or `BaseImageAugmentation` class and override the `call` method as suggested by @Frightera 
- Do you just want to generate more data using augmentation layers and test your model, not necessarily want it as final model behavior? - in this case you can add augmentation layers outside your model during inference and feed that input to your model.

**Comment by innat** [comment]
@divyashreepathihalli 
I was trying to achieve this for Test-Time-Augmentation ([TTA](https://www.nature.com/articles/s41598-020-61808-3)). (with model (defined by any API, sequential, functional or subclass).

**Comment by divyashreepathihalli** [comment]
In that case I would recommend option one specified above.

**Comment by innat** [comment]
@divyashreepathihalli 
How it can be controlled in a more elegant way with a same model?

```
model = [InputLayer, Preprocessing, Augmentation, ...]
model.fit <- Augment
model.evaluate <- No Augment
model.predict <- Augment (for TTA)
```
To make TTA in the prediction time, it is needed to rebuild the augmentation layer (with `training=True`) of the trained model. In that case, I like to test it how it goes with all modeling API (seq, func, sub). The possible problem might be the insert and remove layer in some cases. 

---

Please, don't close the issue w/o asking the OP and set `stat:awaiting response from contributor`.
